File: app/src/main/java/com/shuneault/netrunnerdeckbuilder/ChooseIdentityActivity.java
package com.shuneault.netrunnerdeckbuilder;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v4.view.ViewPager.OnPageChangeListener;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemSelectedListener;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.Spinner;
import android.widget.Toast;

import com.shuneault.netrunnerdeckbuilder.adapters.IdentitySpinnerAdapter;
import com.shuneault.netrunnerdeckbuilder.fragments.NewDeckChooseIdentityFragment;
import com.shuneault.netrunnerdeckbuilder.game.Card;
import com.shuneault.netrunnerdeckbuilder.helper.AppManager;
import com.shuneault.netrunnerdeckbuilder.helper.Sorter;

import java.util.ArrayList;
import java.util.Collections;

public class ChooseIdentityActivity extends AppCompatActivity {

    public static final String EXTRA_SIDE_CODE = "com.example.netrunnerdeckbuilder.EXTRA_SIDE_CODE";
    public static final String EXTRA_INITIAL_IDENTITY_CODE = "com.example.netrunnerdeckbuilder.EXTRA_INITIAL_IDENTITY_CODE";

    public static final String EXTRA_IDENTITY_CODE = "com.example.netrunnerdeckbuilder.EXTRA_IDENTITY_CODE";

    private ArrayList<Card> mIdentities;
    private Button btnOK;
    private Button btnCancel;
    private Spinner spinIdentities;
    private ViewPager pagerIdentity;
    private Card mIdentity;
    private String mSideCode;
    private String mInitialIdentity;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.setContentView(R.layout.fragment_pick_identity);

        // GUI
        spinIdentities = (Spinner) findViewById(R.id.spinIdentities);
        btnOK = (Button) findViewById(R.id.btnOK);
        btnCancel = (Button) findViewById(R.id.btnCancel);
        pagerIdentity = (ViewPager) findViewById(R.id.pagerIdentity);

        // Arguments
        mSideCode = getIntent().getStringExtra(EXTRA_SIDE_CODE);
        mInitialIdentity = getIntent().getStringExtra(EXTRA_INITIAL_IDENTITY_CODE);

        // The identities
        mIdentities = AppManager.getInstance().getCardsFromDataPacksToDisplay().getIdentities(mSideCode);
        // Quit if no identities
        if (mIdentities.size() == 0) {
            Toast.makeText(this, "No identities in packs", Toast.LENGTH_LONG).show();
            finish();
            return;
        }

        Collections.sort(mIdentities, new Sorter.IdentitySorter());
        ArrayAdapter<Card> adapterCard = new IdentitySpinnerAdapter(this, mIdentities);
        spinIdentities.setAdapter(adapterCard);

        // Set the view pager adapter
        pagerIdentity.setAdapter(new ImagePagerAdapter(getSupportFragmentManager()));
        mIdentity = mIdentities.get(pagerIdentity.getCurrentItem());

        // Events
        btnOK.setOnClickListener(new OnClickListener() {

            @Override
            public void onClick(View v) {
                Intent data = new Intent();
                data.putExtra(EXTRA_IDENTITY_CODE, mIdentity.getCode());
                setResult(RESULT_OK, data);
                finish();
            }
        });
        btnCancel.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                finish();
            }
        });
        spinIdentities.setOnItemSelectedListener(new OnItemSelectedListener() {

            @Override
            public void onItemSelected(AdapterView<?> arg0, View arg1,
                                       int arg2, long arg3) {
                pagerIdentity.setCurrentItem(arg2, false);
            }

            @Override
            public void onNothingSelected(AdapterView<?> arg0) {

            }
        });
        pagerIdentity.setOnPageChangeListener(new OnPageChangeListener() {

            @Override
            public void onPageSelected(int arg0) {
                mIdentity = mIdentities.get(arg0);
                spinIdentities.setSelection(arg0);
            }

            @Override
            public void onPageScrolled(int arg0, float arg1, int arg2) {
                //

            }

            @Override
            public void onPageScrollStateChanged(int arg0) {
                //

            }
        });

        // Display the initial identity if any
        if (mInitialIdentity != null) {
            for (int i = 0; i < mIdentities.size(); i++) {
                if (mIdentities.get(i).getCode().equals(mInitialIdentity)) {
                    pagerIdentity.setCurrentItem(i);
                    break;
                }
            }
        }

    }

    private class ImagePagerAdapter extends FragmentStatePagerAdapter {

        public ImagePagerAdapter(FragmentManager fm) {
            super(fm);
        }

        @Override
        public int getCount() {
            return mIdentities.size();
        }


        @Override
        public Fragment getItem(int arg0) {
            // Change the label
            NewDeckChooseIdentityFragment x = new NewDeckChooseIdentityFragment();
            Bundle bundle = new Bundle();
            bundle.putString(NewDeckChooseIdentityFragment.ARGUMENT_IDENTITY_CODE, mIdentities.get(arg0).getCode());
            x.setArguments(bundle);
            return x;

        }

    }
}

File: app/src/main/java/com/shuneault/netrunnerdeckbuilder/ChooseIdentityActivity.kt
package com.shuneault.netrunnerdeckbuilder

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.*
import android.widget.AdapterView.OnItemSelectedListener
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import com.shuneault.netrunnerdeckbuilder.adapters.IdentitySpinnerAdapter
import com.shuneault.netrunnerdeckbuilder.db.CardRepository
import com.shuneault.netrunnerdeckbuilder.fragments.NewDeckChooseIdentityFragment
import com.shuneault.netrunnerdeckbuilder.game.Card
import com.shuneault.netrunnerdeckbuilder.game.Format
import com.shuneault.netrunnerdeckbuilder.helper.AppManager
import com.shuneault.netrunnerdeckbuilder.helper.Sorter.IdentitySorter
import org.koin.android.ext.android.inject
import java.util.*

class ChooseIdentityActivity : AppCompatActivity() {
    private var mIdentities: ArrayList<Card> = ArrayList<Card>()
    private var btnOK: Button? = null
    private var btnCancel: Button? = null
    private var spinIdentities: Spinner? = null
    private var pagerIdentity: ViewPager? = null
    private var mIdentity: Card? = null
    private var mSideCode: String? = null
    private var mInitialIdentity: String? = null
    private var mFormat: Format? = null

    val cardRepo: CardRepository by inject()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        this.setContentView(R.layout.fragment_pick_identity)

        // GUI
        spinIdentities = findViewById<View>(R.id.spinIdentities) as Spinner
        btnOK = findViewById<View>(R.id.btnOK) as Button
        btnCancel = findViewById<View>(R.id.btnCancel) as Button
        pagerIdentity = findViewById<View>(R.id.pagerIdentity) as ViewPager
        val repo = cardRepo

        // Arguments
        val intent = intent
        mSideCode = intent.getStringExtra(EXTRA_SIDE_CODE)
        mInitialIdentity = intent.getStringExtra(EXTRA_INITIAL_IDENTITY_CODE)
        val formatID = intent.getIntExtra(EXTRA_FORMAT, 0)
        mFormat = if (formatID > 0) repo.getFormat(formatID) else repo.defaultFormat

        // The identities
        mIdentities = repo.getCardPool(mFormat).getIdentities(mSideCode)
        // Quit if no identities
        if (mIdentities.size == 0) {
            Toast.makeText(this, "No identities in packs", Toast.LENGTH_LONG).show()
            finish()
            return
        }
        Collections.sort(mIdentities, IdentitySorter())
        val adapterCard: ArrayAdapter<Card> = IdentitySpinnerAdapter(this, mIdentities)
        spinIdentities!!.adapter = adapterCard

        // Set the view pager adapter
        pagerIdentity!!.adapter = ImagePagerAdapter(supportFragmentManager)
        mIdentity = mIdentities.get(pagerIdentity!!.currentItem)

        // Events
        btnOK!!.setOnClickListener {
            val data = Intent()
            data.putExtra(EXTRA_IDENTITY_CODE, mIdentity!!.getCode())
            setResult(RESULT_OK, data)
            finish()
        }
        btnCancel!!.setOnClickListener { finish() }
        spinIdentities!!.onItemSelectedListener = object : OnItemSelectedListener {
            override fun onItemSelected(
                arg0: AdapterView<*>?, arg1: View,
                arg2: Int, arg3: Long
            ) {
                pagerIdentity!!.setCurrentItem(arg2, false)
            }

            override fun onNothingSelected(arg0: AdapterView<*>?) {}
        }
        pagerIdentity!!.setOnPageChangeListener(object : OnPageChangeListener {
            override fun onPageSelected(arg0: Int) {
                mIdentity = mIdentities.get(arg0)
                spinIdentities!!.setSelection(arg0)
            }

            override fun onPageScrolled(arg0: Int, arg1: Float, arg2: Int) {
                //
            }

            override fun onPageScrollStateChanged(arg0: Int) {
                //
            }
        })

        // Display the initial identity if any
        if (mInitialIdentity != null) {
            for (i in mIdentities.indices) {
                if (mIdentities[i].code == mInitialIdentity) {
                    pagerIdentity!!.currentItem = i
                    break
                }
            }
        }
    }

    private inner class ImagePagerAdapter(fm: FragmentManager?) : FragmentStatePagerAdapter(
        fm!!
    ) {
        override fun getCount(): Int {
            return mIdentities.size
        }

        override fun getItem(arg0: Int): Fragment {
            // Change the label
            val x = NewDeckChooseIdentityFragment()
            val bundle = Bundle()
            bundle.putString(
                NewDeckChooseIdentityFragment.ARGUMENT_IDENTITY_CODE,
                mIdentities[arg0].code
            )
            x.arguments = bundle
            return x
        }
    }

    companion object {
        const val EXTRA_SIDE_CODE = "com.example.netrunnerdeckbuilder.EXTRA_SIDE_CODE"
        const val EXTRA_FORMAT = "EXTRA_FORMAT"
        const val EXTRA_INITIAL_IDENTITY_CODE =
            "com.example.netrunnerdeckbuilder.EXTRA_INITIAL_IDENTITY_CODE"
        const val EXTRA_IDENTITY_CODE = "com.example.netrunnerdeckbuilder.EXTRA_IDENTITY_CODE"
    }
}